### DISCUSSION TITLE
Sharing scripts between PCs

### DISCUSSION THREAD

**Discussion by maraf** [gitextensions/gitextensions#10612] [Q&A] [answered]
Is there a way to share scripts between PCs?

**Comment by vbjay** [comment, marked as answer]
No.  Manually copying correct part from settings file and manually updating the scripts is the only way now. 

My suggestion.... Use gists.github.com and update GitHub plugin to add a scripts gist url in settings and then provide gist syncing in scripts settings.

**Reply by maraf** [reply]
That is actually an interesting idea. Thank you!

**Reply by vbjay** [reply]
Sure. Only concern I see is potential for hardcoded paths and such. As long as variables are used. Making sure programs are in path. So you don't sync a script that references a different path on a different computer.

**Reply by vbjay** [reply]
Even better if you can create multi file gists so that the gist contains 1 file per script.  Could store all script details as json or xml in the gist files.

**Reply by maraf** [reply]
> Sure. Only concern I see is potential for hardcoded paths and such. As long as variables are used. Making sure programs are in path. So you don't sync a script that references a different path on a different computer. 

Yeah, that makes sense. In my case, all my scripts are using programs included in PATH.

> Even better if you can create multi file gists so that the gist contains 1 file per script. Could store all script details as json or xml in the gist files.

Having a single gist and script per file sounds ideal. I would prefere not to encode content as JSON/XML, but then there isn't a place for metadata.

**Reply by vbjay** [reply]
Right. Exactly why you encode as json or xml as the content in the gist file. Same as how we encode it in xml in the settings. So the gist file name could be guid while options like icon, name, run background and such would be key/value pairs in xml or json.

Gist file <some guid>
```json
{
"Name":"Fetch After Push",
"RunBackground":false,...
}

```
Then next gist file has other script and so on. 

Any change to scripts just delete all files and add back from local using same guids of existing and commit a gist update. Then sync is just a fetch and reload. 

Bonus gists are repos so how was this script configured 3 versions ago.. easy...  So script versioning could also be implemented.

**Reply by maraf** [reply]
Okay, your idea growth on me over night and I really like it now!

**Reply by vbjay** [reply]
So the reson for gist multi files is so user could hit push or pull in script settings for one script and pull down or update the one script or push or pull whole script collection. Also just to be safe is add a do not sync checkbox to scripts to prevent a script from being included.

**Reply by vbjay** [reply]
Oh one more thought. Add a file import/export that replicates the gist file load/save so users can load or export to a file selecting which scripts they want to load. Just in case they have an offline computer they have to sneaker-net. Just my thoughts... I'd write the scripts to temp and then create a zip file so if user wanted to inspect they could by opening in zip program.

**Comment by RussKie** [comment]
I have a POC in my fork where scripts could be saved in the repo's config,
but it's not 100% complete.

**Reply by RussKie** [reply]
Here it is: https://github.com/RussKie/gitextensions/tree/scripts_serialize

**Reply by maraf** [reply]
That sounds good, but most of my scripts are "global" and syncing them between machines hurt.

**Reply by RussKie** [reply]
Some of my scripts are repo specific, and this would really help with this use case.

@vbjay's idea with installable scripts is quite cool too.

**Reply by vbjay** [reply]
@RussKie  https://github.com/gitextensions/gitextensions/discussions/10612#discussioncomment-4611614 and the one after that should cover your concerns.
